doc/warm-reboot/view_switch.md

```diff
@@ -0,0 +1,179 @@
+upon which all the objects in new view will get synced to consumer.
+
+However, the actual operations get passed to consumer should be optimized as best effort, 
+so if an object is present in both old view and new view, 
+it should not be `del()` and `set()` again.
+
+In order to archieve that, we are going to add two additional APIs: `create_temp_view()` and `apply_temp_view()`,
+as well as modifying existing `del()` and `set()` API.
+
+This design is based on the assumption that there is only one producer writing to a specific table. 
+If there are multiple producer writing to a same table, the changes that 'secondary' producer write to the table
+while 'main' producer was doing view switching might be lost.
+
+
+### Redis Objects and Memory Objects
+As a reference, here are the exiting Redis objects corresponding to a certain table:
+
+| Terminology   | Sample Redis Object  |
+| ------------- | :-----------------:  |
+| TableHash     | ` ROUTE_TABLE:25.78.106.0/27` |
+| StateHash     | `_ROUTE_TABLE:25.78.106.0/27` |
+| KeySet        | `ROUTE_TABLE_KEY_SET`      |
+| DelKeySet     | `ROUTE_TABLE_DEL_SET`      |
+| Channel       | `ROUTE_TABLE_CHANNEL`      |
+
+We are not going to create additional Redis object to support the temporary view. 
+Instead, an in-memory table `TableDump m_tempViewState` 
+that shares a similar structure with TableHash and StateHash will be created to maintain the target state of the temporary view.
+Additionally, a flag `bool m_tempViewActive` will indicates whether a temporary view is current being worked on or not.
+
+### Operation Pseudo Code
+
+#### `create_temp_view()`
+
+Simply mark `m_tempViewActive` flag as active and initialize `m_tempViewState`.
+```
+m_tempViewActive = true;
+m_tempViewState.clear(); 
+```
+
+#### `set()` and `del()`
+
+When `m_tempViewActive == false`, current behavior of `set()` and `del()` won't be modified.
+
+When `m_tempViewActive == true`, `set()` and `del()` will only modify temporary view content in `m_tempViewState` instead of writing into DB.
+
+`set()`:
+
+```
+if (m_tempViewActive)
+{
+    for (const auto& iv: values)
+    {
+        m_tempViewState['key'][fvField(iv)] = fvValue(iv);
+    }
+}
+else 
+{
+    # Current behavior
+    SADD KeySet key
+    for field, value do
+        HSET StateHash:key field value
+    end
+    PUBLISH Channel
+}
+```
+
+`del()`:
+
+```
+if (m_tempViewActive)
+{
+    m_tempViewState.erase('key');
+}
+else 
+{
+    # Current behavior
+    SADD KeySet key
+    SADD DelKeySet key
+    DEL StateHash:key
+    PUBLISH Channel
+}
+```
+
+#### `apply_temp_view()`
+
+To apply temporary view, there are several steps:
+1. Drop all pending operations by clearing KeySet, DelKeySet, and StateHash
+2. Dump content of current view
+3. Compare current view and target view, generate a minimal set of `set()` and `del()` operations
+4. Write to KeySet, DelKeySet, and StateHash accordingly, publish the change to the channel, and unset `m_tempViewActive` flag.
```

Comment:
what about missing field? can we just delete the field instead of delete the key and and the object?

Reply:
It's not supported even in the normal 'SET'/'DEL' operations currently. What's the scenario that new object will have a missing field, while we cannot just delete the object and create a new one?